### PR TITLE
refactor of read/write nodes in dht_live_nodes_alert to reusable functions

### PR DESCRIPTION
**Pull request by aldenml**
I will need these functions in https://github.com/arvidn/libtorrent/pull/1652

### CODE REVIEW

File: src/alert.cpp [hunk 1]
```diff
@@ -2136,28 +2136,32 @@ namespace libtorrent
...
 			else
-				m_v6_num_nodes++;
+				v6_num_nodes++;
 #endif
 		}
 
-		m_v4_nodes_idx = alloc.allocate(m_v4_num_nodes * (20 + 6));
-		m_v6_nodes_idx = alloc.allocate(m_v6_num_nodes * (20 + 18));
+		aux::allocation_slot const v4_nodes_idx = alloc.allocate(v4_num_nodes * (20 + 6));
+		aux::allocation_slot const v6_nodes_idx = alloc.allocate(v6_num_nodes * (20 + 18));
```
Comment (arvidn) [inline src/alert.cpp]: does it make sense to allocate space for IPv6 nodes if ``TORRENT_USE_IPV6`` is not set?
I wonder if this was a deliberate decision or an oversight

File: src/alert.cpp [hunk 2]
```diff
@@ -2136,28 +2136,32 @@ namespace libtorrent
...
 		, std::vector<std::pair<sha1_hash, udp::endpoint>> const& nodes)
-		: node_id(nid)
-		, m_alloc(alloc)
 	{
+		int v4_num_nodes = 0;
+		int v6_num_nodes = 0;
+
 		for (auto const& n : nodes)
 		{
 			if (n.second.protocol() == udp::v4())
```
Comment (arvidn) [inline src/alert.cpp]: if you invert this check you can avoid it entirely when IPv6 is disabled

File: src/alert.cpp (hunk 1 above)
Comment (aldenml) [inline src/alert.cpp]: it was deliberate, and it's `0`, no actual allocation

File: src/alert.cpp (hunk 2 above)
Comment (aldenml) [inline src/alert.cpp]: jeje, this is intentional, a safety feature, since it's `udp::endpoint`

**Comment by arvidn** [comment]
iiuc, you'r just factoring out the work previously done in the constructor to a separate function, as well as the work in assembling the node vector. is that right?

**Comment by aldenml** [comment]
That's correct, because I have plans to use these functions in my other PR (https://github.com/arvidn/libtorrent/pull/1652) to return `nodes`.

**Pull request merged**